### PR TITLE
Refund-api [master] Add refund feature to afterpay API

### PR DESCRIPTION
**Pull request by rajneeshsharma9**

### CODE REVIEW

File: lib/afterpay/refund.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+          requestId: payment_id,
+          merchantRefernce: reference,
+          amount: Utils::Money.api_hash(amount)
+        }
+      end
+
+      new(request.body)
+    end
+
+    attr_accessor :id, :status, :created, :total, :order, :error
```
Comment (sawangupta92) [inline lib/afterpay/refund.rb]: move these to the top of the file

File: lib/afterpay/refund.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    def self.execute(order_id:, reference:, payment_id:, amount:)
+      request = Afterpay.client.post("/v1/payments/#{ order_id }/refund") do |req|
+        req.body = {
+          requestId: payment_id,
+          merchantRefernce: reference,
+          amount: Utils::Money.api_hash(amount)
+        }
+      end
+
+      new(request.body)
```
Comment (sawangupta92) [inline lib/afterpay/refund.rb:29]: let me know the list of attributes & a sample response

File: lib/afterpay/client.rb [hunk 3]
```diff
@@ -43,6 +43,9 @@ def default_connection
         Faraday.new(url: self.class.server_url) do |conn|
           conn.use ErrorMiddleware if Afterpay.config.raise_errors
           conn.authorization "Basic", self.class.auth_token
+          # Afterpay requires User-Agent header to be set for all requests
+          # Refer https://docs.afterpay.com/au-online-api-v1.html#configuration
+          conn.headers['User-Agent'] = Afterpay.config.user_agent_header if Afterpay.config.user_agent_header.present?
```
Comment (sawangupta92) [inline lib/afterpay/client.rb]: why do we need this?

File: lib/afterpay/refund.rb (hunk 1 above)
Comment (sawangupta92) [inline lib/afterpay/refund.rb]: where are you using all these attributes? Please omit the ones you are not using

File: lib/afterpay/refund.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+module Afterpay
+  # The Refund object
+  class Refund
+    # Refunds the Payment
+    # @param order_id [String] unique order id returned by afterpay when capturing payment
+    # @param reference [String] the reference for order at merchant end
+    # @param payment_id [String] Unique Payment id at merchant end
+    # @param amount [Money] amount to be refunded
+    # @return [Refund] the Refund object
```
Comment (sawangupta92) [inline lib/afterpay/refund.rb:12]: document the accessible attributes in the refund object as well

File: lib/afterpay/refund.rb (hunk 1 above)
Comment (sawangupta92) [inline lib/afterpay/refund.rb]: this should not be attr_accessor, why are we allowing it to be updated from outside

**Review by sawangupta92 — CHANGES_REQUESTED** [review]

File: lib/afterpay/refund.rb (hunk 2 above)
Comment (rajneeshsharma9) [reply lib/afterpay/refund.rb:29]: Please refer https://docs.afterpay.com/au-online-api-v1.html#create-refund

File: lib/afterpay/client.rb (hunk 3 above)
Comment (rajneeshsharma9) [reply lib/afterpay/client.rb]: This is not a feature in gem already but afterpay requires this header present in all requests sent to them, hence we are adding feature to intialize this.

Comment (sawangupta92) [reply lib/afterpay/client.rb]: This should be a separate PR then

File: lib/afterpay/refund.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+          amount: Utils::Money.api_hash(amount)
+        }
+      end
+
+      new(request.body)
+    end
+
+    # Initialize Refund from response
+    def initialize(attributes)
+      @refund_id = attributes[:refundId]
```
Comment (sawangupta92) [inline lib/afterpay/refund.rb:34]: let's also initialize amount Object & merchantReference

File: lib/afterpay/refund.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+module Afterpay
+  # The Refund object
+  class Refund
+    # refund_id [string]: unique ID returned by afterpay after successful refund
```
Comment (sawangupta92) [inline lib/afterpay/refund.rb]: move these lines after Line 15, also add another line like this

# == Response [Refund] Object
# refund_id [string]: unique ID returned by afterpay after successful refund
# refunded_at [timestamp]: The timestamp at which the refund was done
# error [hash]: error information if refund fails

File: lib/afterpay/refund.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    # refunded_at [timestamp]: The timestamp at which the refund was done
+    # error [hash]: error information if refund fails
+    attr_reader :refund_id, :refunded_at, :error
```
Comment (sawangupta92) [inline lib/afterpay/refund.rb]: move this out of comment area

**Review by sawangupta92 — CHANGES_REQUESTED** [review]

**Comment by sawangupta92** [comment]
Also, make sure to push everything in one single commit

**Pull request merged**